aws/resource_aws_internet_gateway.go

```diff
@@ -63,9 +63,14 @@ func resourceAwsInternetGatewayCreate(d *schema.ResourceData, meta interface{})
 			return resource.NonRetryableError(err)
 		}
 	})
-
+	if isResourceTimeoutError(err) {
+		igRaw, _, err = IGStateRefreshFunc(conn, d.Id())()
+		if igRaw == nil {
+			return fmt.Errorf("Error refreshing internet gateway state: nil state received")
```

Comment:
As an operator, I'm thinking this error message might not be so obvious what to do -- I'd suggest changing this to be more indicative of what happened from their perspective so they know how to remediate in the situation, e.g.

```suggestion
			return fmt.Errorf("error finding Internet Gateway (%s) after creation; retry running Terraform", d.Id())
```

Also the retry code above is fairly confusing as well as the logic currently reads:

- if we found the Internet Gateway, good
- if we received no error (and no Internet Gateway either since we are past the above conditional), good (`return resource.RetryableError(err)` is `return resource.RetryableError(nil)` which really is `return nil` -- see also https://github.com/hashicorp/terraform/pull/17220)
- we received an error, bad

It might make more sense for the original logic to also actually retry on a similar error message, e.g.

```go
	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
		igRaw, _, err = IGStateRefreshFunc(conn, d.Id())()

		if err != nil {
			return resource.NonRetryableError(err)
		}

		if igRaw == nil {
			return resource.RetryableError(fmt.Errorf("error finding Internet Gateway (%s) after creation; retry running Terraform", d.Id()))
		}

		return nil
	})
```

That said, this retry logic may not solve the real intention here of handling EC2 eventual consistency issues after creation for _later_ API calls. This retry logic may return success while later API calls still fail. Given the current resource code, typically we would prefer to handle the eventual consistency by wrapping `setTags` and `resourceAwsInternetGatewayAttach` with `resource.Retry()` instead of a separate `resource.Retry()` before those calls. Its probably not worth changing this all right now though.